subsys/bluetooth/services/ots.c

```diff
@@ -0,0 +1,385 @@
+
+LOG_MODULE_REGISTER(bt_gatt_ots, CONFIG_BT_GATT_OTS_LOG_LEVEL);
+
+static ssize_t ots_feature_read(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr, void *buf,
+				uint16_t len, uint16_t offset)
+{
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) attr->user_data;
+
+	LOG_DBG("OTS Feature GATT Read Operation");
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &ots->features,
+		sizeof(ots->features));
+}
+
+static ssize_t ots_obj_name_read(struct bt_conn *conn,
+				 const struct bt_gatt_attr *attr, void *buf,
+				 uint16_t len, uint16_t offset)
+{
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) attr->user_data;
+
+	LOG_DBG("OTS Object Name GATT Read Operation");
+
+	if (!ots->cur_obj) {
+		LOG_DBG("No Current Object selected in OTS!");
+		return BT_GATT_ERR(BT_GATT_OTS_OBJECT_NOT_SELECTED);
+	}
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset,
+				 ots->cur_obj->metadata.name,
+				 strlen(ots->cur_obj->metadata.name) + 1);
+}
+
+static ssize_t ots_obj_type_read(struct bt_conn *conn,
+				 const struct bt_gatt_attr *attr, void *buf,
+				 uint16_t len, uint16_t offset)
+{
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) attr->user_data;
+	struct bt_gatt_ots_obj_metadata *obj_meta;
+
+	LOG_DBG("OTS Object Type GATT Read Operation");
+
+	if (!ots->cur_obj) {
+		LOG_DBG("No Current Object selected in OTS!");
+		return BT_GATT_ERR(BT_GATT_OTS_OBJECT_NOT_SELECTED);
+	}
+
+	obj_meta = &ots->cur_obj->metadata;
+	if (obj_meta->type.uuid.type == BT_UUID_TYPE_128) {
+		return bt_gatt_attr_read(conn, attr, buf, len, offset,
+					 obj_meta->type.uuid_128.val,
+					 sizeof(obj_meta->type.uuid_128.val));
+	} else {
+		return bt_gatt_attr_read(conn, attr, buf, len, offset,
+					 &obj_meta->type.uuid_16.val,
+					 sizeof(obj_meta->type.uuid_16.val));
+	}
+}
+
+static ssize_t ots_obj_size_read(struct bt_conn *conn,
+				 const struct bt_gatt_attr *attr, void *buf,
+				 uint16_t len, uint16_t offset)
+{
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) attr->user_data;
+
+	LOG_DBG("OTS Object Size GATT Read Operation");
+
+	if (!ots->cur_obj) {
+		LOG_DBG("No Current Object selected in OTS!");
+		return BT_GATT_ERR(BT_GATT_OTS_OBJECT_NOT_SELECTED);
+	}
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset,
+				 &ots->cur_obj->metadata.size,
+				 sizeof(ots->cur_obj->metadata.size));
+}
+
+static ssize_t ots_obj_id_read(struct bt_conn *conn,
+			       const struct bt_gatt_attr *attr, void *buf,
+			       uint16_t len, uint16_t offset)
+{
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) attr->user_data;
+	struct bt_gatt_ots_obj_id id;
+
+	LOG_DBG("OTS Object ID GATT Read Operation");
+
+	if (!ots->cur_obj) {
+		LOG_DBG("No Current Object selected in OTS!");
+		return BT_GATT_ERR(BT_GATT_OTS_OBJECT_NOT_SELECTED);
+	}
+
+	id.val = ots->cur_obj->metadata.id;
+	LOG_HEXDUMP_DBG(id.ba, sizeof(id.ba), "Current Object ID:");
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, id.ba,
+				 sizeof(id.ba));
```

Comment:
I think you meant sys_put

Reply:
No, I did mean `sys_get_le48`, as that would get the lower 48 bits of the uint64 ID and then return them, to be used as the value for bt_gatt_attr_read. In other words, it would look like

```
return bt_gatt_attr_read(conn, attr, buf, len, offset, sys_get_le48(id), 6);
```

Reply:
oh, ok. I would like to use sys_put for log_hexdump (getting rid of this ID struct).